File: lib/src/models/pitch.dart
import 'package:meta/meta.dart';
import 'package:musicxml_parser/src/exceptions/musicxml_structure_exception.dart';
import 'package:musicxml_parser/src/exceptions/musicxml_validation_exception.dart';
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:musicxml_parser/src/utils/validation_utils.dart';
import 'package:xml/xml.dart';

/// Represents a musical pitch, including its step, octave, and alteration.
///
/// A pitch is defined by a step (A-G), an octave (0-9), and an optional
/// alteration (e.g., sharp, flat). This class ensures that pitch values
/// conform to standard musical notation.
///
/// Objects of this class are immutable.
@immutable
class Pitch {
  /// The musical step of the pitch, represented as a capital letter (C, D, E, F, G, A, B).
  final String step;

  /// The octave number (0-9) in which the pitch resides.
  final int octave;

  /// The chromatic alteration of the pitch.
  ///
  /// Positive values represent sharps (e.g., 1 for a single sharp),
  /// negative values represent flats (e.g., -1 for a single flat),
  /// and 0 or null represents no alteration. Typically ranges from -2 to 2.
  final int? alter;

  /// Creates a new [Pitch] instance.
  ///
  /// The constructor itself does basic assertions if any, but robust validation
  /// (e.g., ensuring step is a valid letter, octave is in range) is typically
  /// handled by factory constructors like [Pitch.fromXmlElement] or validation utilities.
  const Pitch({
    required this.step,
    required this.octave,
    this.alter,
  });

  /// Creates a new [Pitch] instance from an MusicXML `<pitch>` [element].
  ///
  /// This factory parses the required `<step>` and `<octave>` elements,
  /// and the optional `<alter>` element. It performs validation against
  /// MusicXML rules (e.g., valid pitch steps, octave range, alter range).
  ///
  /// Throws [MusicXmlStructureException] if required XML elements are missing.
  /// Throws [MusicXmlValidationException] if the parsed values are invalid
  /// according to MusicXML specifications.
  ///
  /// Parameters:
  ///   [element]: The XML element representing the `<pitch>`.
  ///   [partId]: The ID of the part, used for context in error messages.
  ///   [measureNumber]: The number of the measure, used for context in error messages.
  factory Pitch.fromXmlElement(
    XmlElement element,
    String partId,
    String measureNumber,
  ) {
    final line = XmlHelper.getLineNumber(element);

    final stepElement = element.findElements('step').firstOrNull;
    if (stepElement == null) {
      throw MusicXmlStructureException(
        'Required <step> element not found in <pitch>',
        parentElement: 'pitch',
        line: line,
        context: {'part': partId, 'measure': measureNumber},
      );
    }
    final step = stepElement.innerText.trim();

    final octaveElement = element.findElements('octave').firstOrNull;
    if (octaveElement == null) {
      throw MusicXmlStructureException(
        'Required <octave> element not found in <pitch>',
        parentElement: 'pitch',
        line: line,
        context: {'part': partId, 'measure': measureNumber},
      );
    }
    final octaveText = octaveElement.innerText.trim();
    final octave = int.tryParse(octaveText);

    final alterElement = element.findElements('alter').firstOrNull;
    final alterText = alterElement?.innerText.trim();
    final alter = alterText != null ? int.tryParse(alterText) : null;

    // Perform validation after extracting values
    if (!ValidationUtils.validPitchSteps.contains(step)) {
      throw MusicXmlValidationException(
        'Invalid pitch step: "$step". Must be one of ${ValidationUtils.validPitchSteps.join(", ")}.',
        rule: 'pitch_step_invalid',
        line: line,
        context: {'part': partId, 'measure': measureNumber, 'parsedStep': step},
      );
    }

    if (octave == null ||
        octave < ValidationUtils.minOctave ||
        octave > ValidationUtils.maxOctave) {
      throw MusicXmlValidationException(
        'Invalid octave: "$octaveText". Must be an integer between ${ValidationUtils.minOctave} and ${ValidationUtils.maxOctave}.',
        rule: 'pitch_octave_invalid',
        line: line,
        context: {
          'part': partId,
          'measure': measureNumber,
          'parsedOctave': octaveText
        },
      );
    }

    if (alterText != null && (alter == null || alter < -2 || alter > 2)) {
      throw MusicXmlValidationException(
        'Invalid alter value: "$alterText". If present, must be an integer between -2 and 2.',
        rule: 'pitch_alter_invalid',
        line: line,
        context: {
          'part': partId,
          'measure': measureNumber,
          'parsedAlter': alterText
        },
      );
    }

    return Pitch(step: step, octave: octave, alter: alter);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Pitch &&
          runtimeType == other.runtimeType &&
          step == other.step &&
          octave == other.octave &&
          alter == other.alter;

  @override
  int get hashCode => step.hashCode ^ octave.hashCode ^ (alter?.hashCode ?? 0);

  @override
  String toString() => 'Pitch{step: $step, octave: $octave, alter: $alter}';
}

File: csharp/MusicXMLParser/Models/Pitch.cs
using System;
using System.Xml.Linq;
using System.Linq;
using MusicXMLParser.Exceptions; // Assuming you will create this namespace for exceptions
using MusicXMLParser.Utils; // Assuming you will create this namespace for utils

namespace MusicXMLParser.Models
{
    /// <summary>
    /// Represents a musical pitch, including its step, octave, and alteration.
    /// </summary>
    /// <remarks>
    /// A pitch is defined by a step (A-G), an octave (0-9), and an optional
    /// alteration (e.g., sharp, flat). This class ensures that pitch values
    /// conform to standard musical notation.
    /// Objects of this class are immutable.
    /// </remarks>
    public class Pitch
    {
        /// <summary>
        /// The musical step of the pitch, represented as a capital letter (C, D, E, F, G, A, B).
        /// </summary>
        public string Step { get; }

        /// <summary>
        /// The octave number (0-9) in which the pitch resides.
        /// </summary>
        public int Octave { get; }

        /// <summary>
        /// The chromatic alteration of the pitch.
        /// </summary>
        /// <remarks>
        /// Positive values represent sharps (e.g., 1 for a single sharp),
        /// negative values represent flats (e.g., -1 for a single flat),
        /// and 0 or null represents no alteration. Typically ranges from -2 to 2.
        /// </remarks>
        public int? Alter { get; }

        /// <summary>
        /// Creates a new <see cref="Pitch"/> instance.
        /// </summary>
        public Pitch(string step, int octave, int? alter = null)
        {
            // Basic validation can be done here if desired, though more complex validation
            // is often handled by factory methods or dedicated validation utilities.
            if (string.IsNullOrWhiteSpace(step))
                throw new ArgumentException("Step cannot be null or empty.", nameof(step));
            if (octave < ValidationUtils.MinOctave || octave > ValidationUtils.MaxOctave) // Assuming ValidationUtils will be ported
                throw new ArgumentOutOfRangeException(nameof(octave), $"Octave must be between {ValidationUtils.MinOctave} and {ValidationUtils.MaxOctave}.");
            if (alter.HasValue && (alter.Value < -2 || alter.Value > 2)) // Example range
                throw new ArgumentOutOfRangeException(nameof(alter), "Alter, if specified, must be between -2 and 2.");


            Step = step;
            Octave = octave;
            Alter = alter;
        }

        /// <summary>
        /// Creates a new <see cref="Pitch"/> instance from an MusicXML &lt;pitch&gt; <paramref name="element"/>.
        /// </summary>
        /// <remarks>
        /// This factory parses the required &lt;step&gt; and &lt;octave&gt; elements,
        /// and the optional &lt;alter&gt; element. It performs validation against
        /// MusicXML rules (e.g., valid pitch steps, octave range, alter range).
        /// Throws <see cref="MusicXmlStructureException"/> if required XML elements are missing.
        /// Throws <see cref="MusicXmlValidationException"/> if the parsed values are invalid
        /// according to MusicXML specifications.
        /// </remarks>
        /// <param name="element">The XML element representing the &lt;pitch&gt;.</param>
        /// <param name="partId">The ID of the part, used for context in error messages.</param>
        /// <param name="measureNumber">The number of the measure, used for context in error messages.</param>
        public static Pitch FromXElement(XElement element, string partId, string measureNumber)
        {
            var line = element.Attribute("line")?.Value; // Or however line numbers are accessed

            var stepElement = element.Element("step");
            if (stepElement == null)
            {
                throw new MusicXmlStructureException(
                    "Required <step> element not found in <pitch>",
                    parentElement: "pitch",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber } }
                );
            }
            var step = stepElement.Value.Trim();

            var octaveElement = element.Element("octave");
            if (octaveElement == null)
            {
                throw new MusicXmlStructureException(
                    "Required <octave> element not found in <pitch>",
                    parentElement: "pitch",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber } }
                );
            }
            var octaveText = octaveElement.Value.Trim();
            if (!int.TryParse(octaveText, out var octave))
            {
                 throw new MusicXmlValidationException(
                    $"Invalid octave: \"{octaveText}\". Must be an integer.",
                    rule: "pitch_octave_invalid_format",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber }, {"parsedOctave", octaveText} }
                );
            }


            var alterElement = element.Element("alter");
            var alterText = alterElement?.Value.Trim();
            int? alter = null;
            if (!string.IsNullOrEmpty(alterText))
            {
                if (!int.TryParse(alterText, out var parsedAlter))
                {
                    throw new MusicXmlValidationException(
                        $"Invalid alter value: \"{alterText}\". If present, must be an integer.",
                        rule: "pitch_alter_invalid_format",
                        line: line,
                        context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber }, {"parsedAlter", alterText} }
                    );
                }
                alter = parsedAlter;
            }

            // Perform validation after extracting values
            if (!ValidationUtils.ValidPitchSteps.Contains(step))
            {
                throw new MusicXmlValidationException(
                    $"Invalid pitch step: \"{step}\". Must be one of {string.Join(", ", ValidationUtils.ValidPitchSteps)}.",
                    rule: "pitch_step_invalid",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber }, { "parsedStep", step } }
                );
            }

            if (octave < ValidationUtils.MinOctave || octave > ValidationUtils.MaxOctave)
            {
                throw new MusicXmlValidationException(
                    $"Invalid octave: \"{octaveText}\". Must be an integer between {ValidationUtils.MinOctave} and {ValidationUtils.MaxOctave}.",
                    rule: "pitch_octave_invalid_range",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber }, { "parsedOctave", octaveText } }
                );
            }

            if (alter.HasValue && (alter.Value < -2 || alter.Value > 2)) // Example range, align with ValidationUtils if it has specific constants
            {
                throw new MusicXmlValidationException(
                    $"Invalid alter value: \"{alterText}\". If present, must be an integer between -2 and 2.",
                    rule: "pitch_alter_invalid_range",
                    line: line,
                    context: new Dictionary<string, string> { { "part", partId }, { "measure", measureNumber }, { "parsedAlter", alterText } }
                );
            }

            return new Pitch(step, octave, alter);
        }


        public override bool Equals(object obj)
        {
            if (obj is Pitch other)
            {
                return Step == other.Step &&
                       Octave == other.Octave &&
                       Alter == other.Alter;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Step, Octave, Alter);
        }

        public override string ToString() => $"Pitch{{Step: {Step}, Octave: {Octave}, Alter: {Alter?.ToString() ?? "null"}}}";
    }
}
